c_src/js.cc

```diff
@@ -73,19 +76,116 @@ create_jscontext(size_t max_mem)
 }
 
 
-JSCx::JSCx(size_t max_mem) : _cx(create_jscontext(max_mem), JS_DestroyContext)
+bool
+check_timeout_shim(JSContext* jscx)
 {
+    JSCx* cx = (JSCx*) JS_GetContextPrivate(jscx);
+    return cx->check_timeout();
+}
+
+
+JSCx::JSCx(size_t max_mem) : _cx(create_jscontext(max_mem), JS_DestroyContext),
+        _wd_alive(true), _wd_active(false)
+{
+    JS_SetContextPrivate(this->_cx.get(), this);
+    JS_AddInterruptCallback(this->_cx.get(), &check_timeout_shim);
+
+    this->_wd_alive = true;
+    this->_wd_active = false;
+    this->_wd_thread = std::make_unique<std::thread>(&JSCx::wd_run, this);
+}
+
+
+JSCx::~JSCx()
+{
+    // Scope the lock acquisition
+    {
+        std::unique_lock<std::mutex> guard(this->_wd_lock);
+        this->_wd_alive = false;
+        this->_wd_cv.notify_one();
+    }
+    this->_wd_thread->join();
 }
 
 
 std::unique_ptr<JSCompartment>
 JSCx::new_compartment()
 {
-    return std::make_unique<JSCompartment>(this->_cx.get());
+    return std::make_unique<JSCompartment>(this, this->_cx.get());
 }
 
 
-JSCompartment::JSCompartment(JSContext* cx) : _cx(cx)
+void
+JSCx::set_timeout(int ms_timeout)
+{
+    int us_timeout = ms_timeout * 1000;
+    auto now = std::chrono::system_clock::now();
+    this->_deadline = now + std::chrono::system_clock::duration(us_timeout);
+    this->_timed_out = false;
+}
+
+
+bool
+JSCx::check_timeout()
+{
+    if(std::chrono::system_clock::now() > this->_deadline) {
+        this->_timed_out = true;
+        return false;
+    }
+
+    return true;
+}
+
+
+bool
+JSCx::timed_out()
+{
+    return this->_timed_out;
+}
+
+
+void
+JSCx::set_watchdog_status(bool status)
+{
+    std::unique_lock<std::mutex> guard(this->_wd_lock);
+    this->_wd_active = status;
+    this->_wd_cv.notify_one();
+}
+
+
+void
+JSCx::wd_run()
+{
+    std::unique_lock<std::mutex> guard(this->_wd_lock);
+    std::chrono::system_clock::duration sleep_time;
+    while(this->_wd_alive) {
+        JS_RequestInterruptCallback(this->_cx.get());
+
+        if(this->_wd_active) {
+            sleep_time = std::chrono::system_clock::duration(ACTIVE_SLEEP_TIME);
```

Comment:
Do we have to explicitly cast duration to microseconds here, or is it guaranteed to be microseconds for most practical architecture?

Reply:
Good catch... I went to link to where the `system_clock` was defined as microseconds but then realized that's not true at all. I must've just looked at a value on my machine locally and deduced that.

So, given that the system clock doesn't come with a guaranteed resolution the appropriate fix is to cast from a duration with a known precision to the system_clock duration at runtime. I've updated the PR to do just that. The cast implementation is in the new `wait_time` function which is then re-used anywhere we need a system_clock duration which includes the `MAX_SLEEP_TIME` you caught above. Now the max sleep is a bit shy of 25 days.